src/lightly_train/_data/yolo_object_detection_dataset.py

```diff
@@ -0,0 +1,165 @@
+#
+# Copyright (c) Lightly AG and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+#
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Literal, Sequence
+
+import numpy as np
+import pydantic
+import torch
+from torch.utils.data import Dataset
+
+from lightly_train._configs.config import PydanticConfig
+from lightly_train._data import file_helpers
+from lightly_train._data.task_data_args import TaskDataArgs
+from lightly_train._transforms.task_transform import TaskTransform
+from lightly_train.types import ImageFilename, ObjectDetectionDatasetItem, PathLike
+
+
+class YOLOObjectDetectionDataset(Dataset[ObjectDetectionDatasetItem]):
+    def __init__(
+        self,
+        dataset_args: YOLOObjectDetectionDatasetArgs,
+        image_filenames: Sequence[ImageFilename],
+        transform: TaskTransform,
+    ) -> None:
+        self.args = dataset_args
+        self.image_filenames = image_filenames
+        self.transform = transform
+
+        self._image_dir, self._label_dir = (
+            dataset_args.image_dir,
+            dataset_args.label_dir,
+        )
+
+    def __len__(self) -> int:
+        return len(self.image_filenames)
+
+    def __getitem__(self, index: int) -> ObjectDetectionDatasetItem:
+        assert self._image_dir is not None
+        assert self._label_dir is not None
```

Comment:
Nit: This can't ever be `None` because `YOLOObjectDetectionDatasetArgs .image_dir` and `.label_dir` are typed as `Path`.  I would also recommend to use `self.args.image_dir` and `self.args.label_dir` directly instead of storing the same information twice in the class.